src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs

```diff
@@ -350,28 +349,34 @@ await LogTestHostCreatedAsync(
             return toolsTestHost;
         }
 
-        NamedPipeClient? dotnetTestPipeClient = await ConnectToDotnetTestPipeIfAvailableAsync(commandLineHandler, testApplicationCancellationTokenSource);
+        DotnetTestConnection dotnetTestConnection = new(serviceProvider, testApplicationCancellationTokenSource);
+        bool isConnectedToDotnetTest = await dotnetTestConnection.TryConnectToDotnetTestPipeIfAvailableAsync();
+
+        if (isConnectedToDotnetTest)
+        {
+            serviceProvider.AddService(dotnetTestConnection);
```

Comment:
I think we can always add it and retrive where is needed, we can check there `IsConnected` as condition to "do something or not" like register or not the custom `IDataConsumer`

Reply:
Never mind me if you want to use `? GetService()` but I don't want to register the `IDataCollector` if we're not connected.